xonsh/ptk/key_bindings.py

```diff
@@ -41,3 +41,33 @@ def insert_literal_tab(event):
         """
         event.cli.current_buffer.insert_text(env.get('INDENT'))
 
+    @handle(Keys.ControlJ, filter=IsMultiline())
+    def multiline_carriage_return(event):
+        """
+        Preliminary parser to determine if 'Enter' key should send command to
+        the xonsh parser for execution or should insert a newline for continued
+        input.
+
+        Current 'triggers' for inserting a newline are:
+        - Not on first line of buffer and line is non-empty
+        - Previous character is a colon (covers if, for, etc...)
+        - User is in an open paren-block
+        - Line ends with backslash
+        - Any text exists below cursor position (relevant when editing previous
+        multiline blocks)
+        """
+        b = event.cli.current_buffer
+        if (not b.document.on_first_line and
+           not b.document.current_line.isspace()):
+            b.newline(copy_margin=True)
+        elif b.document.char_before_cursor == ':':
+            b.newline()
+            b.insert_text(env.get('INDENT'), fire_event=False)
+        elif b.document.text.count('(') > b.document.text.count(')'):
```

Comment:
I don't think that this is safe on its own. What if someone has a string with a single open parentheses, like `print('(')`.  BaseShell() will set the `self.need_more_lines` boolean attribute in the event of SyntaxError on `push()`.  However, I don't think that the prompt-toolkit method uses the `push()` method.  Also the `push()` method is for gluing together independent lines, so is probably not appropriate here.

Rather, you probably should do what the `push()` method does and see it the current code can be parsed, [like here](https://github.com/scopatz/xonsh/blob/master/xonsh/base_shell.py#L165).  Namely,

```python
mode = 'single' if b.document.text.count('\n') == 0 else 'exec'
try:
      code = self.execer.compile(b.document.text, mode=mode,
                                       glbs=None,
                                       locs=self.ctx)
except SyntaxError:
   ...
```